cephfs/path_test.go

```diff
@@ -60,3 +60,273 @@ func TestRemoveDir(t *testing.T) {
+
+		err = mount.Unlink(dirname)
+		// Error, not permitted on directory.
+		assert.Error(t, err)
+	})
+}
+
+func TestSymlink(t *testing.T) {
+	mount := fsConnect(t)
+	defer fsDisconnect(t, mount)
+
+	t.Run("file operations - source doesn't exist", func(t *testing.T) {
+		fname1 := "TestFile1.txt"
+
+		err := mount.Symlink(fname1, "Symlnk1")
+		// No Error, symlink works even if source file doesn't exist.
+		assert.NoError(t, err)
+		_, err = os.Stat(path.Join(CephMountDir, "Symlnk1"))
+		// Error, source is not there.
+		assert.Error(t, err)
+
+		localPath := path.Join(CephMountDir, fname1)
+		_, err = os.Stat(localPath)
+		// Error, source file is still not there.
+		assert.EqualError(t, err,
+			fmt.Sprintf("stat %s: no such file or directory", localPath))
+
+		f1, err := mount.Open(fname1, os.O_RDWR|os.O_CREATE, 0666)
+		assert.NoError(t, err)
+		assert.NotNil(t, f1)
+		defer func() {
+			assert.NoError(t, f1.Close())
+			assert.NoError(t, mount.Unlink(fname1))
+		}()
+		err = mount.Symlink(fname1, "Symlnk1")
+		// Error, Symlink1 exists.
+		assert.Error(t, err)
+		defer func() {
+			assert.NoError(t, mount.Unlink("Symlnk1"))
+		}()
+	})
+
+	t.Run("file operation - source exists", func(t *testing.T) {
+		fname := "TestFile2.txt"
+		f1, err := mount.Open(fname, os.O_RDWR|os.O_CREATE, 0666)
+		assert.NoError(t, err)
+		assert.NotNil(t, f1)
+		defer func() {
+			assert.NoError(t, f1.Close())
+			assert.NoError(t, mount.Unlink(fname))
+		}()
+		err = mount.Symlink(fname, "Symlnk2")
+		// No Error, normal symlink operation.
+		assert.NoError(t, err)
+		defer func() {
+			assert.NoError(t, mount.Unlink("Symlnk2"))
+		}()
+		_, err = os.Stat(path.Join(CephMountDir, "Symlnk2"))
+		assert.NoError(t, err)
+	})
+
+	t.Run("directory operation", func(t *testing.T) {
+		err := mount.Symlink("/", "/")
+		assert.Error(t, err)
+
+		err = mount.Symlink("/", "someDir")
+		assert.NoError(t, err)
+
+		err = mount.Symlink("someFile", "/")
+		// Error, permission denied.
+		assert.Error(t, err)
+
+		// 1. Create a directory.
+		// 2. Create a symlink to that directory.
+		// 3. Create a file inside symlink.
+		// 4. Ensure that it is not a directory.
+		dirname := "mydir"
+		err = mount.MakeDir(dirname, 0755)
+		assert.NoError(t, err)
+		defer func() {
+			assert.NoError(t, mount.RemoveDir(dirname))
+		}()
+
+		err = mount.Symlink(dirname, "symlnk")
+		assert.NoError(t, err)
+		defer func() {
+			assert.NoError(t, mount.Unlink("symlnk"))
+		}()
+
+		fname := "symlnk/file"
+		f1, err := mount.Open(fname, os.O_RDWR|os.O_CREATE, 0666)
+		assert.NoError(t, err)
+		assert.NotNil(t, f1)
+		defer func() {
+			assert.NoError(t, f1.Close())
+			assert.NoError(t, mount.Unlink(fname))
+		}()
+		var fileInfo os.FileInfo
+		fileInfo, err = os.Stat(path.Join(CephMountDir, "symlnk/file"))
+		assert.NoError(t, err)
+		assert.Equal(t, fileInfo.IsDir(), false)
+	})
+}
+
+func TestReadlink(t *testing.T) {
+	mount := fsConnect(t)
+	defer fsDisconnect(t, mount)
+
+	fname := "file1.txt"
```

Comment:
Like the tests above you ought to break this up into sub-tests for each condition you are checking, conflict on wrong file type, non existing path, happy-path success, etc.

Reply:
Done